internal/encryption/encryption.go

```diff
@@ -16,30 +18,91 @@ import (
 	"github.com/tink-crypto/tink-go/v2/tink"
 )
 
-type SubKey string
+type Subkey interface {
+	Salt() string
+}
 
-const (
-	TimelineSubKey SubKey = "timeline"
-	AsyncSubKey    SubKey = "async"
-)
+type TimelineSubkey struct{}
+
+func (t TimelineSubkey) Salt() string {
+	return "timeline"
+}
+
+type AsyncSubkey struct{}
+
+func (a AsyncSubkey) Salt() string {
+	return "async"
+}
+
+type KeyStoreProvider interface {
+	// EnsureKey asks a provider to check for an encrypted key.
+	// If not available, call the generateKey function to create a new key.
+	// The provider should handle transactions around checking and setting the key, to prevent race conditions.
+	EnsureKey(ctx context.Context, generateKey func() ([]byte, error)) ([]byte, error)
+}
+
+// Builder constructs a DataEncryptor when used with a provider.
+// Use a chain of With* methods to configure the builder.
+type Builder struct {
+	kmsURI      optional.Option[string]
+	awsV1Client *awsv1kms.KMS
+}
+
+func NewBuilder() *Builder {
+	return &Builder{
+		kmsURI: optional.None[string](),
+	}
+}
+
+// WithKMSURI sets the URI for the KMS key to use. Omitting this call or using nil will create a NoOpEncryptor.
+func (b *Builder) WithKMSURI(kmsURI *string) *Builder {
+	b.kmsURI = optional.Ptr[string](kmsURI)
+	return b
+}
+
+func (b *Builder) WithAWSV1Client(kms *awsv1kms.KMS) *Builder {
+	b.awsV1Client = kms
```

Comment:
This doesn't seem to be used, can it be deleted?

Reply:
Good catch. I thought I needed it for one of the tests but turns out it got bypassed. YAGNI.